sweep/fee_bumper.go

```diff
@@ -0,0 +1,623 @@
+
+	// Height specifies the height at which the request was made. This
+	// should be viewed as the input's sweeping transaction's broadcast
+	// height.
+	Height uint32
+
+	// ConfTarget is the sweep transaction's confirmation target.
+	ConfTarget uint32
+
+	// FeeRate is the requested fee rate when the sweeping transaction is
+	// created.
+	FeeRate chainfee.SatPerKWeight
+
+	// MaxFeeRatio specifies a portion of the input amount that can be used
+	// as mining fee. If not set, defaultMaxFeeRatio will be used.
+	MaxFeeRatio float64
+
+	// Strategy specifies how the fee will be bumped.
+	Strategy Strategy
+}
+
+// String gives a verbose string which can be used in logging.
+func (m *MonitorRequest) String() string {
+	s := fmt.Sprintf("outpoint %v: strategy=%s, fee rate=%v",
+		m.Input.OutPoint(), m.Strategy, m.FeeRate)
+
+	if m.ConfTarget != NoHeightInfo {
+		s += fmt.Sprintf(", conf target=%v", m.ConfTarget)
+	}
+
+	if m.MaxFeeRatio != 0 {
+		s += fmt.Sprintf(", max fee ratio=%v", m.MaxFeeRatio)
+	} else {
+		s += fmt.Sprintf(", max fee ratio=%v", defaultMaxFeeRatio)
+	}
+
+	return s
+}
+
+// estimateMaxFeeRate gives an estimated max fee rate based on the requested
+// MaxFeeRatio. It assumes the input will be swept inside a transition with one
+// input and one change output, and calcualtes the fee rate using
+// (max fee) / (transaction weight). The changeOutputScript is specified when
+// creating the fee bumper, so it's shared across all inputs. Since the actual
+// sweeping transaction may have multiple inputs and one change output, the
+// actual fee used will be smaller than the max allowed fee.
+//
+// TODO(yy): allow the weight or max fee rate to be specified via the request?
+func (m *MonitorRequest) estimateMaxFeeRate(
+	changeOutputScript []byte) (chainfee.SatPerKWeight, error) {
+
+	// Get the max fee ratio. If not set, use the default.
+	maxFeeRatio := m.MaxFeeRatio
+	if maxFeeRatio == 0 {
+		maxFeeRatio = defaultMaxFeeRatio
+	}
+
+	// Calculate the max fee.
+	amount := btcutil.Amount(m.Input.SignDesc().Output.Value)
+	maxFee := amount.MulF64(maxFeeRatio)
+
+	// Create a new weight estimator. The fee rate we pass in here doesn't
+	// matter as we only use it for calculating the weight.
+	weightEstimate := newWeightEstimator(m.FeeRate)
+
+	// Add the change output.
+	switch {
+	case txscript.IsPayToTaproot(changeOutputScript):
+		weightEstimate.addP2TROutput()
+
+	case txscript.IsPayToWitnessScriptHash(changeOutputScript):
+		weightEstimate.addP2WSHOutput()
+
+	case txscript.IsPayToWitnessPubKeyHash(changeOutputScript):
+		weightEstimate.addP2WKHOutput()
+
+	case txscript.IsPayToPubKeyHash(changeOutputScript):
+		weightEstimate.estimator.AddP2PKHOutput()
+
+	case txscript.IsPayToScriptHash(changeOutputScript):
+		weightEstimate.estimator.AddP2SHOutput()
+
+	default:
+		// Unknown script type.
+		return 0, errors.New("unknown script type")
+	}
+
+	// Add the input.
+	if err := weightEstimate.add(m.Input); err != nil {
```

Comment:
This doesn't seem to factor in the additional overhead of the type of inputs _added_ to bump the transaction.